File: 0036-valid-sudoku/36.go
package main

func box(row, col int) int {

	return (row/3)*3 + col/3

}

func isValidSudoku(board [][]byte) bool {

	cols := make(map[int]map[byte]bool)
	rows := make(map[int]map[byte]bool)
	boxes := make(map[int]map[byte]bool)

	for i := 0; i < 9; i++ {
		cols[i] = make(map[byte]bool)
		rows[i] = make(map[byte]bool)
		boxes[i] = make(map[byte]bool)
	}

	for row := 0; row < 9; row++ {

		for col := 0; col < 9; col++ {

			item := board[row][col]

			if item == '.' {
				continue
			}

			index := box(row, col)

			box := boxes[index]
			row := rows[row]
			col := cols[col]

			if box[item] || row[item] || col[item] {
				return false
			}

			box[item] = true
			row[item] = true
			col[item] = true

		}
	}

	return true
}

File: 0036-valid-sudoku/36.cpp
#include <set>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
  int block(int row, int col) { return (row / 3) * 3 + col / 3; }

  bool isValidSudoku(vector<vector<char>> &board) {
    unordered_map<int, unordered_map<char, bool>> cols, rows, blocks;

    for (int i = 0; i < 9; i++) {
      for (int j = 0; j < 9; j++) {

        int cell = board[i][j];

        if (cell == '.') {
          continue;
        }

        int k = block(i, j);

        bool row = rows[i][cell], col = cols[j][cell], block = blocks[k][cell];

        if (block || row || col) {
          return false;
        }

        rows[i][cell] = true;
        cols[j][cell] = true;
        blocks[k][cell] = true;
      }
    }

    return true;
  }
};